query/hash_reduction.cu

```diff
@@ -0,0 +1,395 @@
+  // hash_value. Therefore the first index paired with the hash value will
+  // be the final output value.
+  template<typename map_type>
+  void reduce(map_type *map,
+              DimensionVector inputKeys,
+              uint8_t *inputValues,
+              cudaStream_t cudaStream) {
+    DimensionHashIterator<32,
+                          thrust::counting_iterator<uint32_t>>
+        hashIter(inputKeys.DimValues,
+                 inputKeys.NumDimsPerDimWidth,
+                 inputKeys.VectorCapacity);
+
+    auto rawMapKeyIter = thrust::make_zip_iterator(
+        thrust::make_tuple(hashIter, thrust::counting_iterator<uint32_t>(0)));
+
+    auto hashIndexFusionFunc = []
+    __host__ __device__(
+        typename decltype(rawMapKeyIter)::value_type tuple) {
+      return (static_cast<int64_t>(thrust::get<0>(tuple)) << 32)
+          | (static_cast<int64_t>(thrust::get<1>(tuple)));
+    };
+
+    auto mapKeyIter = thrust::make_transform_iterator(rawMapKeyIter,
+                                                      hashIndexFusionFunc);
+
+    auto mapKeyValuePairIter = thrust::make_zip_iterator(
+        thrust::make_tuple(
+            mapKeyIter,
+            reinterpret_cast<value_type *>(inputValues)));
+
+    auto equality = []
+    __host__ __device__(key_type lhs, key_type rhs) {
+      return lhs >> 32 == rhs >> 32;
+    };
+
+    auto insertionFunc = [=]
+    __host__ __device__(
+        thrust::tuple<key_type, value_type> key_value_tuple) {
+      map->insert(tuple2pair(key_value_tuple), f, equality);
+    };
+
+    thrust::for_each_n(GET_EXECUTION_POLICY(cudaStream),
+                       mapKeyValuePairIter,
+                       length,
+                       insertionFunc);
+  }
+
+  template<typename map_type>
+  int output(map_type *map,
+             DimensionVector inputKeys,
+             DimensionVector outputKeys, uint8_t *outputValue,
+             cudaStream_t cudaStream) {
+    // calculate prefix sum of NumDimsPerDimWidth so that it will be easier
+    // to tell the valueBytes given a dim index. The prefix sum is inclusive.
+    uint8_t
+    dimWidthPrefixSum[NUM_DIM_WIDTH];
+    for (int i = 0; i < NUM_DIM_WIDTH; i++) {
+      dimWidthPrefixSum[i] = inputKeys.NumDimsPerDimWidth[i];
+      if (i > 0) {
+        dimWidthPrefixSum[i] += dimWidthPrefixSum[i - 1];
+      }
+    }
+
+    device_unique_ptr<uint32_t> globalWriteIndexDevice =
+        make_device_unique<uint32_t>(cudaStream);
+
+    ExtractGroupByResultFunctor<map_type> extractorFunc(
+        map, inputKeys.DimValues, inputKeys.VectorCapacity,
+        outputKeys.DimValues, dimWidthPrefixSum,
+        outputValue, globalWriteIndexDevice.get());
+
+    thrust::for_each_n(GET_EXECUTION_POLICY(cudaStream),
+                       thrust::counting_iterator<uint32_t>(0),
+                       length, extractorFunc);
+    // copy the reduced count back from GPU.
+    uint32_t
+    globalWriteIndexHost;
+    ares::asyncCopyDeviceToHost(
```

Comment:
async + global seems fishy..

Reply:
yeah, you are right, we should use the synced copy here

Reply:
or better to use cudaWaitForStream